<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Crystal Docs 1.13.3">
<meta name="crystal_docs.project_version" content="master">
<meta name="crystal_docs.project_name" content="chem">



<link href="css/style.css" rel="stylesheet" type="text/css" />
<script type="text/javascript" src="js/doc.js"></script>

  <meta name="repository-name" content="chem">
  <title>chem master</title>
  <script type="text/javascript">
  CrystalDocs.base_path = "";
  </script>
</head>
<body>

<svg class="hidden">
  <symbol id="octicon-link" viewBox="0 0 16 16">
    <path fill="currentColor" fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
  </symbol>
</svg>
<input type="checkbox" id="sidebar-btn">
<label for="sidebar-btn" id="sidebar-btn-label">
  <svg class="open" xmlns="http://www.w3.org/2000/svg" height="2em" width="2em" viewBox="0 0 512 512"><title>Open Sidebar</title><path fill="currentColor" d="M80 96v64h352V96H80zm0 112v64h352v-64H80zm0 112v64h352v-64H80z"></path></svg>
  <svg class="close" xmlns="http://www.w3.org/2000/svg" width="2em" height="2em" viewBox="0 0 512 512"><title>Close Sidebar</title><path fill="currentColor" d="m118.6 73.4-45.2 45.2L210.7 256 73.4 393.4l45.2 45.2L256 301.3l137.4 137.3 45.2-45.2L301.3 256l137.3-137.4-45.2-45.2L256 210.7Z"></path></svg>
</label>
<div class="sidebar">
  <div class="sidebar-header">
    <div class="search-box">
      <input type="search" class="search-input" placeholder="Search..." spellcheck="false" aria-label="Search">
    </div>

    <div class="project-summary">
      <h1 class="project-name">
        <a href="index.html">
          chem
        </a>
      </h1>

      <span class="project-version">
        master
      </span>
    </div>
  </div>

  <div class="search-results hidden">
    <ul class="search-list"></ul>
  </div>

  <div class="types-list">
    <ul>
  
  <li class=" " data-id="chem/Array" data-name="array(t)">
      <a href="Array.html">Array</a>
      
    </li>
  
  <li class=" " data-id="chem/Char" data-name="char">
      <a href="Char.html">Char</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem" data-name="chem">
      <a href="Chem.html">Chem</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Angle" data-name="chem::angle">
      <a href="Chem/Angle.html">Angle</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Atom" data-name="chem::atom">
      <a href="Chem/Atom.html">Atom</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/AtomContainer" data-name="chem::atomcontainer">
      <a href="Chem/AtomContainer.html">AtomContainer</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/AtomView" data-name="chem::atomview">
      <a href="Chem/AtomView.html">AtomView</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Bias" data-name="chem::bias">
      <a href="Chem/Bias.html">Bias</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Bond" data-name="chem::bond">
      <a href="Chem/Bond.html">Bond</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/BondArray" data-name="chem::bondarray">
      <a href="Chem/BondArray.html">BondArray</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/BondOrder" data-name="chem::bondorder">
      <a href="Chem/BondOrder.html">BondOrder</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Chain" data-name="chem::chain">
      <a href="Chem/Chain.html">Chain</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/ChainView" data-name="chem::chainview">
      <a href="Chem/ChainView.html">ChainView</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Connectivity" data-name="chem::connectivity(t)">
      <a href="Chem/Connectivity.html">Connectivity</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/Cube" data-name="chem::cube">
      <a href="Chem/Cube.html">Cube</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Cube/Reader" data-name="chem::cube::reader">
      <a href="Chem/Cube/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Cube/Writer" data-name="chem::cube::writer">
      <a href="Chem/Cube/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/DCD" data-name="chem::dcd">
      <a href="Chem/DCD.html">DCD</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/DCD/Reader" data-name="chem::dcd::reader">
      <a href="Chem/DCD/Reader.html">Reader</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/Dihedral" data-name="chem::dihedral">
      <a href="Chem/Dihedral.html">Dihedral</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/DX" data-name="chem::dx">
      <a href="Chem/DX.html">DX</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/DX/Reader" data-name="chem::dx::reader">
      <a href="Chem/DX/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/DX/Writer" data-name="chem::dx::writer">
      <a href="Chem/DX/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/Element" data-name="chem::element">
      <a href="Chem/Element.html">Element</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Error" data-name="chem::error">
      <a href="Chem/Error.html">Error</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Format" data-name="chem::format">
      <a href="Chem/Format.html">Format</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/FormatReader" data-name="chem::formatreader(t)">
      <a href="Chem/FormatReader.html">FormatReader</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/FormatReader/Attached" data-name="chem::formatreader::attached(t)">
      <a href="Chem/FormatReader/Attached.html">Attached</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/FormatReader/Headed" data-name="chem::formatreader::headed(t)">
      <a href="Chem/FormatReader/Headed.html">Headed</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/FormatReader/Indexable" data-name="chem::formatreader::indexable(t)">
      <a href="Chem/FormatReader/Indexable.html">Indexable</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/FormatReader/MultiEntry" data-name="chem::formatreader::multientry(t)">
      <a href="Chem/FormatReader/MultiEntry.html">MultiEntry</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/FormatWriter" data-name="chem::formatwriter(t)">
      <a href="Chem/FormatWriter.html">FormatWriter</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/FormatWriter/MultiEntry" data-name="chem::formatwriter::multientry(t)">
      <a href="Chem/FormatWriter/MultiEntry.html">MultiEntry</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/Gen" data-name="chem::gen">
      <a href="Chem/Gen.html">Gen</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Gen/Reader" data-name="chem::gen::reader">
      <a href="Chem/Gen/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Gen/Writer" data-name="chem::gen::writer">
      <a href="Chem/Gen/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/Improper" data-name="chem::improper">
      <a href="Chem/Improper.html">Improper</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/JDFTx" data-name="chem::jdftx">
      <a href="Chem/JDFTx.html">JDFTx</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/JDFTx/Reader" data-name="chem::jdftx::reader">
      <a href="Chem/JDFTx/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/JDFTx/Writer" data-name="chem::jdftx::writer">
      <a href="Chem/JDFTx/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/Metadata" data-name="chem::metadata">
      <a href="Chem/Metadata.html">Metadata</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Metadata/Any" data-name="chem::metadata::any">
      <a href="Chem/Metadata/Any.html">Any</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Metadata/ValueType" data-name="chem::metadata::valuetype">
      <a href="Chem/Metadata/ValueType.html">ValueType</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/Mol" data-name="chem::mol">
      <a href="Chem/Mol.html">Mol</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Mol/Reader" data-name="chem::mol::reader">
      <a href="Chem/Mol/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Mol/Variant" data-name="chem::mol::variant">
      <a href="Chem/Mol/Variant.html">Variant</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Mol/Writer" data-name="chem::mol::writer">
      <a href="Chem/Mol/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/Mol2" data-name="chem::mol2">
      <a href="Chem/Mol2.html">Mol2</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Mol2/Reader" data-name="chem::mol2::reader">
      <a href="Chem/Mol2/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Mol2/Writer" data-name="chem::mol2::writer">
      <a href="Chem/Mol2/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/ParseException" data-name="chem::parseexception">
      <a href="Chem/ParseException.html">ParseException</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/PDB" data-name="chem::pdb">
      <a href="Chem/PDB.html">PDB</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/PDB/Hybrid36" data-name="chem::pdb::hybrid36">
      <a href="Chem/PDB/Hybrid36.html">Hybrid36</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/PDB/Reader" data-name="chem::pdb::reader">
      <a href="Chem/PDB/Reader.html">Reader</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/PDB/Writer" data-name="chem::pdb::writer">
      <a href="Chem/PDB/Writer.html">Writer</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/PDB/Writer/BondOptions" data-name="chem::pdb::writer::bondoptions">
      <a href="Chem/PDB/Writer/BondOptions.html">BondOptions</a>
      
    </li>
  
</ul>

      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/PeriodicTable" data-name="chem::periodictable">
      <a href="Chem/PeriodicTable.html">PeriodicTable</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/Protein" data-name="chem::protein">
      <a href="Chem/Protein.html">Protein</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Protein/DSSP" data-name="chem::protein::dssp">
      <a href="Chem/Protein/DSSP.html">DSSP</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Protein/HlxParams" data-name="chem::protein::hlxparams">
      <a href="Chem/Protein/HlxParams.html">HlxParams</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Protein/PSIQUE" data-name="chem::protein::psique">
      <a href="Chem/Protein/PSIQUE.html">PSIQUE</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Protein/SecondaryStructure" data-name="chem::protein::secondarystructure">
      <a href="Chem/Protein/SecondaryStructure.html">SecondaryStructure</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Protein/SecondaryStructureType" data-name="chem::protein::secondarystructuretype">
      <a href="Chem/Protein/SecondaryStructureType.html">SecondaryStructureType</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/Protein/Stride" data-name="chem::protein::stride">
      <a href="Chem/Protein/Stride.html">Stride</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Protein/Stride/Writer" data-name="chem::protein::stride::writer">
      <a href="Chem/Protein/Stride/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/PSF" data-name="chem::psf">
      <a href="Chem/PSF.html">PSF</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/PSF/Reader" data-name="chem::psf::reader">
      <a href="Chem/PSF/Reader.html">Reader</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/PullParser" data-name="chem::pullparser">
      <a href="Chem/PullParser.html">PullParser</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/PyMOL" data-name="chem::pymol">
      <a href="Chem/PyMOL.html">PyMOL</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/PyMOL/Writer" data-name="chem::pymol::writer">
      <a href="Chem/PyMOL/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/RegisterFormat" data-name="chem::registerformat">
      <a href="Chem/RegisterFormat.html">RegisterFormat</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Residue" data-name="chem::residue">
      <a href="Chem/Residue.html">Residue</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/ResidueType" data-name="chem::residuetype">
      <a href="Chem/ResidueType.html">ResidueType</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/ResidueView" data-name="chem::residueview">
      <a href="Chem/ResidueView.html">ResidueView</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/SDF" data-name="chem::sdf">
      <a href="Chem/SDF.html">SDF</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/SDF/Reader" data-name="chem::sdf::reader">
      <a href="Chem/SDF/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/SDF/Writer" data-name="chem::sdf::writer">
      <a href="Chem/SDF/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/Spatial" data-name="chem::spatial">
      <a href="Chem/Spatial.html">Spatial</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Spatial/CoordinatesProxy" data-name="chem::spatial::coordinatesproxy">
      <a href="Chem/Spatial/CoordinatesProxy.html">CoordinatesProxy</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Direction" data-name="chem::spatial::direction">
      <a href="Chem/Spatial/Direction.html">Direction</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Error" data-name="chem::spatial::error">
      <a href="Chem/Spatial/Error.html">Error</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/FloatTriple" data-name="chem::spatial::floattriple">
      <a href="Chem/Spatial/FloatTriple.html">FloatTriple</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/Spatial/Grid" data-name="chem::spatial::grid">
      <a href="Chem/Spatial/Grid.html">Grid</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Spatial/Grid/Dimensions" data-name="chem::spatial::grid::dimensions">
      <a href="Chem/Spatial/Grid/Dimensions.html">Dimensions</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Grid/Info" data-name="chem::spatial::grid::info">
      <a href="Chem/Spatial/Grid/Info.html">Info</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Grid/Location" data-name="chem::spatial::grid::location">
      <a href="Chem/Spatial/Grid/Location.html">Location</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/KDTree" data-name="chem::spatial::kdtree">
      <a href="Chem/Spatial/KDTree.html">KDTree</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Mat3" data-name="chem::spatial::mat3">
      <a href="Chem/Spatial/Mat3.html">Mat3</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/NotPeriodicError" data-name="chem::spatial::notperiodicerror">
      <a href="Chem/Spatial/NotPeriodicError.html">NotPeriodicError</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/NumberTriple" data-name="chem::spatial::numbertriple">
      <a href="Chem/Spatial/NumberTriple.html">NumberTriple</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Parallelepiped" data-name="chem::spatial::parallelepiped">
      <a href="Chem/Spatial/Parallelepiped.html">Parallelepiped</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/PeriodicKDTree" data-name="chem::spatial::periodickdtree">
      <a href="Chem/Spatial/PeriodicKDTree.html">PeriodicKDTree</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Quat" data-name="chem::spatial::quat">
      <a href="Chem/Spatial/Quat.html">Quat</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Size3" data-name="chem::spatial::size3">
      <a href="Chem/Spatial/Size3.html">Size3</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Transform" data-name="chem::spatial::transform">
      <a href="Chem/Spatial/Transform.html">Transform</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Spatial/Vec3" data-name="chem::spatial::vec3">
      <a href="Chem/Spatial/Vec3.html">Vec3</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/Structure" data-name="chem::structure">
      <a href="Chem/Structure.html">Structure</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Structure/Builder" data-name="chem::structure::builder">
      <a href="Chem/Structure/Builder.html">Builder</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/Structure/Experiment" data-name="chem::structure::experiment">
      <a href="Chem/Structure/Experiment.html">Experiment</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Structure/Experiment/Method" data-name="chem::structure::experiment::method">
      <a href="Chem/Structure/Experiment/Method.html">Method</a>
      
    </li>
  
</ul>

      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/Templates" data-name="chem::templates">
      <a href="Chem/Templates.html">Templates</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Templates/Atom" data-name="chem::templates::atom">
      <a href="Chem/Templates/Atom.html">Atom</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/Bond" data-name="chem::templates::bond">
      <a href="Chem/Templates/Bond.html">Bond</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/Builder" data-name="chem::templates::builder">
      <a href="Chem/Templates/Builder.html">Builder</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/Detector" data-name="chem::templates::detector">
      <a href="Chem/Templates/Detector.html">Detector</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/MatchData" data-name="chem::templates::matchdata">
      <a href="Chem/Templates/MatchData.html">MatchData</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/Registry" data-name="chem::templates::registry">
      <a href="Chem/Templates/Registry.html">Registry</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/Residue" data-name="chem::templates::residue">
      <a href="Chem/Templates/Residue.html">Residue</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/Templates/SpecParser" data-name="chem::templates::specparser">
      <a href="Chem/Templates/SpecParser.html">SpecParser</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/Templates/SpecParser/AtomRecord" data-name="chem::templates::specparser::atomrecord">
      <a href="Chem/Templates/SpecParser/AtomRecord.html">AtomRecord</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/SpecParser/BondRecord" data-name="chem::templates::specparser::bondrecord">
      <a href="Chem/Templates/SpecParser/BondRecord.html">BondRecord</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/SpecParser/ImplicitBondRecord" data-name="chem::templates::specparser::implicitbondrecord">
      <a href="Chem/Templates/SpecParser/ImplicitBondRecord.html">ImplicitBondRecord</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/Templates/Ter" data-name="chem::templates::ter">
      <a href="Chem/Templates/Ter.html">Ter</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/VASP" data-name="chem::vasp">
      <a href="Chem/VASP.html">VASP</a>
      
        <ul>
  
  <li class="parent " data-id="chem/Chem/VASP/Chgcar" data-name="chem::vasp::chgcar">
      <a href="Chem/VASP/Chgcar.html">Chgcar</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/VASP/Chgcar/Reader" data-name="chem::vasp::chgcar::reader">
      <a href="Chem/VASP/Chgcar/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/VASP/Chgcar/Writer" data-name="chem::vasp::chgcar::writer">
      <a href="Chem/VASP/Chgcar/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Chem/VASP/GridReader" data-name="chem::vasp::gridreader">
      <a href="Chem/VASP/GridReader.html">GridReader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/VASP/GridWriter" data-name="chem::vasp::gridwriter">
      <a href="Chem/VASP/GridWriter.html">GridWriter</a>
      
    </li>
  
  <li class="parent " data-id="chem/Chem/VASP/Locpot" data-name="chem::vasp::locpot">
      <a href="Chem/VASP/Locpot.html">Locpot</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/VASP/Locpot/Reader" data-name="chem::vasp::locpot::reader">
      <a href="Chem/VASP/Locpot/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/VASP/Locpot/Writer" data-name="chem::vasp::locpot::writer">
      <a href="Chem/VASP/Locpot/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/VASP/Poscar" data-name="chem::vasp::poscar">
      <a href="Chem/VASP/Poscar.html">Poscar</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/VASP/Poscar/Reader" data-name="chem::vasp::poscar::reader">
      <a href="Chem/VASP/Poscar/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/VASP/Poscar/Writer" data-name="chem::vasp::poscar::writer">
      <a href="Chem/VASP/Poscar/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/VMD" data-name="chem::vmd">
      <a href="Chem/VMD.html">VMD</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/VMD/Writer" data-name="chem::vmd::writer">
      <a href="Chem/VMD/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Chem/XYZ" data-name="chem::xyz">
      <a href="Chem/XYZ.html">XYZ</a>
      
        <ul>
  
  <li class=" " data-id="chem/Chem/XYZ/Reader" data-name="chem::xyz::reader">
      <a href="Chem/XYZ/Reader.html">Reader</a>
      
    </li>
  
  <li class=" " data-id="chem/Chem/XYZ/Writer" data-name="chem::xyz::writer">
      <a href="Chem/XYZ/Writer.html">Writer</a>
      
    </li>
  
</ul>

      
    </li>
  
</ul>

      
    </li>
  
  <li class="parent " data-id="chem/Colorize" data-name="colorize">
      <a href="Colorize.html">Colorize</a>
      
        <ul>
  
  <li class=" " data-id="chem/Colorize/ColorRGB" data-name="colorize::colorrgb">
      <a href="Colorize/ColorRGB.html">ColorRGB</a>
      
    </li>
  
</ul>

      
    </li>
  
  <li class=" " data-id="chem/Enumerable" data-name="enumerable(t)">
      <a href="Enumerable.html">Enumerable</a>
      
    </li>
  
  <li class=" " data-id="chem/Float" data-name="float">
      <a href="Float.html">Float</a>
      
    </li>
  
  <li class=" " data-id="chem/Indexable" data-name="indexable(t)">
      <a href="Indexable.html">Indexable</a>
      
    </li>
  
  <li class=" " data-id="chem/Int" data-name="int">
      <a href="Int.html">Int</a>
      
    </li>
  
  <li class=" " data-id="chem/Number" data-name="number">
      <a href="Number.html">Number</a>
      
    </li>
  
  <li class=" " data-id="chem/Path" data-name="path">
      <a href="Path.html">Path</a>
      
    </li>
  
  <li class=" " data-id="chem/Range" data-name="range(b, e)">
      <a href="Range.html">Range</a>
      
    </li>
  
  <li class=" " data-id="chem/Regex" data-name="regex">
      <a href="Regex.html">Regex</a>
      
    </li>
  
  <li class=" " data-id="chem/String" data-name="string">
      <a href="String.html">String</a>
      
    </li>
  
</ul>

  </div>
</div>


<div class="main-content">
<h1><a id="chem.cr" class="anchor" href="#chem.cr">  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>chem.cr</h1>
<p><a href="https://github.com/franciscoadasme/chem.cr/actions/workflows/crystal.yml"><img src="https://github.com/franciscoadasme/chem.cr/actions/workflows/crystal.yml/badge.svg" alt="Crystal CI" /></a>
<a href="https://github.com/franciscoadasme/chem.cr/releases/latest"><img src="https://img.shields.io/github/v/release/franciscoadasme/chem.cr.svg?label=version" alt="Version" /></a>
<a href="https://github.com/franciscoadasme/chem.cr/blob/master/LICENSE"><img src="https://img.shields.io/github/license/franciscoadasme/chem.cr.svg" alt="License" /></a></p>
<p><a href="#features">Features</a> |
<a href="#installation">Installation</a> |
<a href="#usage">Usage</a> |
<a href="#benchmark">Benchmark</a> |
<a href="#roadmap">Roadmap</a> |
<a href="#similar-software">Similar software</a> |
<a href="#testing">Testing</a> |
<a href="#contributing">Contributing</a> |
<a href="#contributors">Contributors</a> |
<a href="#license">License</a></p>
<p>A modern library written in [Crystal]<a href="https://crystal-lang.org">1</a> for manipulating molecular files used in computational chemistry and biology.
It aims to be both fast and easy to use.</p>
<blockquote>
<p>[!NOTE]
PSIQUE was moved to the <a href="https://github.com/franciscoadasme/psique">psique</a> repository.</p>
</blockquote>
<h2><a id="features" class="anchor" href="#features">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Features</h2>
<ul>
<li>Object-oriented API for accessing and manipulating molecules.
It follows the topology commonly used in [Protein Data Bank]<a href="https://en.wikipedia.org/wiki/Protein_Data_Bank_(file_format)">2</a> (PDB) file format: Structure (or model) → Chain → Residue → Atom.</li>
<li>Type safety (assigning a number as the atom's name will result in an error at compilation time).</li>
<li>Support for periodic molecular structures.</li>
<li>Support for common molecular file formats (PDB, SDF, etc.).</li>
<li>Spatial measurements (distance, RMSD, alignment).</li>
<li>Template-based topology reconstruction.</li>
<li>Volumetric data.</li>
<li>Fast performance (see <a href="#benchmark">Benchmark</a> below).</li>
</ul>
<blockquote>
<p>[!IMPORTANT]
This library is in alpha stage, meaning that there is missing functionality, documentation, etc. and there will be breaking changes.</p>
</blockquote>
<h2><a id="installation" class="anchor" href="#installation">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Installation</h2>
<p>Ensure the Crystal compiler is installed:</p>
<pre><code class="language-console">$ crystal -v
Crystal 1.13.1 [0cef61e51] (2024-07-12)

LLVM: 18.1.6
Default target: x86_64-unknown-linux-gnu</code></pre>
<p>If the command fails, you need to install the crystal compiler by
following [these steps]<a href="https://crystal-lang.org/install">3</a>.</p>
<p>Crystal requires listing the dependencies in the <code>shard.yml</code> file.
Let's create a new project:</p>
<pre><code class="language-console">$ crystal init app myapp
    create  myapp/.gitignore
    create  myapp/.editorconfig
    create  myapp/LICENSE
    create  myapp/README.md
    create  myapp/.travis.yml
    create  myapp/shard.yml
    create  myapp/src/myapp.cr
    create  myapp/src/myapp/version.cr
    create  myapp/spec/spec_helper.cr
    create  myapp/spec/myapp_spec.cr
Initialized empty Git repository in /home/crystal/myapp/.git/
$ cd myapp</code></pre>
<p>Add the following to the application's <code>shard.yml</code>:</p>
<pre><code class="language-yaml">dependencies:
  chem:
    github: franciscoadasme/chem.cr</code></pre>
<p>Then, resolve and install missing dependencies:</p>
<pre><code class="language-console">$ shards install</code></pre>
<p>Dependencies are installed into the <code>lib</code> folder.
More about dependencies at the [Requiring files]<a href="https://crystal-lang.org/reference/1.14/syntax_and_semantics/requiring_files.html">4</a> guide.</p>
<h3><a id="requirements" class="anchor" href="#requirements">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Requirements</h3>
<ul>
<li>To run STRIDE analysis, you'll need to set <code>STRIDE_BIN</code> to the STRIDE
executable path.</li>
</ul>
<h2><a id="usage" class="anchor" href="#usage">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Usage</h2>
<p>First require the <code>chem</code> module:</p>
<pre><code class="language-crystal"><span class="k">require</span> <span class="s">&quot;chem&quot;</span></code></pre>
<p>Let's first read a structure:</p>
<pre><code class="language-crystal">struc <span class="o">=</span> <span class="t">Chem</span><span class="t">::</span><span class="t">Structure</span>.read <span class="s">&quot;file.pdb&quot;</span>
struc <span class="c"># =&gt; &lt;Structure &quot;1cbn&quot;: 644 atoms, 47 residues, periodic&gt;</span></code></pre>
<p>You can also use a custom <code>.from_*</code> method to specify reading options when available:</p>
<pre><code class="language-crystal"><span class="t">Chem</span><span class="t">::</span><span class="t">Structure</span>.from_pdb <span class="s">&quot;/path/to/file.pdb&quot;</span>
<span class="t">Chem</span><span class="t">::</span><span class="t">Structure</span>.from_pdb <span class="s">&quot;/path/to/file.pdb&quot;</span>, chains: [<span class="s">&#39;A&#39;</span>] <span class="c"># reads only chain A</span>
<span class="t">Chem</span><span class="t">::</span><span class="t">Structure</span>.from_pdb <span class="s">&quot;/path/to/file.pdb&quot;</span>, het: <span class="n">false</span>    <span class="c"># skips HET atoms</span>
<span class="t">Chem</span><span class="t">::</span><span class="t">Structure</span>.from_pdb <span class="s">&quot;/path/to/file.pdb&quot;</span>, alt_loc: <span class="s">&#39;A&#39;</span>  <span class="c"># selects alternate location A</span></code></pre>
<p>You can access PDB header information via the <code>#experiment</code> property:</p>
<pre><code class="language-crystal"><span class="k">if</span> expt <span class="o">=</span> struc.experiment <span class="c"># checks if experiment data is present</span>
  expt.title           <span class="c"># =&gt; &quot;ATOMIC RESOLUTION (0.83 ANGSTROMS) CRYSTAL STRUCTURE...&quot;</span>
  expt.kind            <span class="c"># =&gt; XRayDiffraction</span>
  expt.resolution      <span class="c"># =&gt; 0.83</span>
  expt.deposition_date <span class="c"># =&gt; 1991-10-11</span>
<span class="k">end</span></code></pre>
<p>You can also read many structures at once:</p>
<pre><code class="language-crystal"><span class="c"># read all models</span>
<span class="t">Array</span>(<span class="t">Chem</span><span class="t">::</span><span class="t">Structure</span>).from_pdb <span class="s">&quot;/path/to/file.pdb&quot;</span>
<span class="c"># read 2th and 5th models</span>
<span class="t">Array</span>(<span class="t">Chem</span><span class="t">::</span><span class="t">Structure</span>).from_pdb <span class="s">&quot;/path/to/file.pdb&quot;</span>, indexes: [<span class="n">1</span>, <span class="n">4</span>]</code></pre>
<p>Alternatively, you could use the reader class directly to read one by one via the <code>#each</code> method:</p>
<pre><code class="language-crystal"><span class="t">Chem</span><span class="t">::</span><span class="t">PDB</span><span class="t">::</span><span class="t">Reader</span>.new(<span class="s">&quot;/path/to/file.pdb&quot;</span>).each { <span class="o">|</span>struc<span class="o">|</span> ... }
<span class="t">Chem</span><span class="t">::</span><span class="t">PDB</span><span class="t">::</span><span class="t">Reader</span>.new(<span class="s">&quot;/path/to/file.pdb&quot;</span>).each(indexes: [<span class="n">1</span>, <span class="n">4</span>]) { <span class="o">|</span>struc<span class="o">|</span> ... }</code></pre>
<h3><a id="topology-access" class="anchor" href="#topology-access">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Topology access</h3>
<p>You can access topology objects using the <code>#dig</code> methods:</p>
<pre><code class="language-crystal">struc.dig(<span class="s">&#39;A&#39;</span>)           <span class="c"># =&gt; &lt;Chain A&gt;</span>
struc.dig(<span class="s">&#39;A&#39;</span>, <span class="n">10</span>)       <span class="c"># =&gt; &lt;Residue A:ARG10&gt;</span>
struc.dig(<span class="s">&#39;A&#39;</span>, <span class="n">10</span>, <span class="s">&quot;CA&quot;</span>) <span class="c"># =&gt; &lt;Atom A:ARG10:CA(146)&gt;</span>

struc.dig <span class="s">&#39;A&#39;</span>, <span class="n">10</span>, <span class="s">&quot;CJ&quot;</span>  <span class="c"># raises a KeyError because &quot;CJ&quot; doesn&#39;t exist</span>
struc.dig? <span class="s">&#39;A&#39;</span>, <span class="n">10</span>, <span class="s">&quot;CJ&quot;</span> <span class="c"># =&gt; nil</span></code></pre>
<p>Each topology object have several modifiable properties:</p>
<pre><code class="language-crystal">atom <span class="o">=</span> struc.dig <span class="s">&#39;A&#39;</span>, <span class="n">10</span>, <span class="s">&quot;CA&quot;</span>
atom.element.name            <span class="c"># =&gt; &quot;Carbon&quot;</span>
atom.coords                  <span class="c"># =&gt; [8.47 4.577 8.764]</span>
atom.occupancy               <span class="c"># =&gt; 1.0</span>
atom.bonded_atoms.map <span class="o">&amp;</span>.name <span class="c"># =&gt; [&quot;N&quot;, &quot;C&quot;, &quot;HA&quot;, &quot;CB&quot;]</span>

atom.residue.number   <span class="c"># =&gt; 10</span>
atom.residue.protein? <span class="c"># =&gt; true</span>
atom.residue.pred     <span class="c"># =&gt; &lt;Residue A:PHE9&gt;</span>

atom.chain.id            <span class="c"># =&gt; &#39;A&#39;</span>
atom.chain.residues.size <span class="c"># =&gt; 152</span></code></pre>
<p>Thanks to Crystal's powerful standard library, manipulating topology objects is very easy:</p>
<pre><code class="language-crystal"><span class="c"># select chains longer than 50 residues</span>
struc.chains.<span class="k">select</span> { <span class="o">|</span>chain<span class="o">|</span> chain.residues.size <span class="o">&gt;</span> <span class="n">50</span> }
<span class="c"># ramachandran angles</span>
struc.residues.map { <span class="o">|</span>residue<span class="o">|</span> {residue.phi, residue.psi} } <span class="c"># =&gt; [{129.5, 90.1}, ...]</span>
<span class="c"># renumber residues starting from 1</span>
struc.residues.each_with_index { <span class="o">|</span>residue, i<span class="o">|</span> residue.number <span class="o">=</span> i <span class="o">+</span> <span class="n">1</span> }
<span class="c"># constrain Z-axis</span>
struc.atoms.each <span class="o">&amp;</span>.constraint<span class="o">=</span>(<span class="n">:z</span>)
<span class="c"># total partial charge</span>
struc.atoms.sum_of <span class="o">&amp;</span>.partial_charge
<span class="c"># iterate over secondary structure elements</span>
struc.residues.chunk(<span class="o">&amp;</span>.sec).each <span class="k">do</span> <span class="o">|</span>sec, residues<span class="o">|</span>
  sec      <span class="c"># =&gt; HelixAlpha</span>
  residues <span class="c"># =&gt; [&lt;Residue A:ARG1&gt;, &lt;Residue A:LEU2&gt;, ...]</span>
<span class="k">end</span></code></pre>
<p>The <code>#chains</code>, <code>#residues</code>, and <code>#atoms</code> methods return an array view of <code>Chain</code>, <code>Residue</code> and <code>Atom</code> instances, respectively.
Refer to the <a href="https://crystal-lang.org/api/latest/Enumerable.html">Enumerable</a> and <a href="https://crystal-lang.org/api/latest/Indexable.html">Indexable</a> modules for more information about available methods.</p>
<h3><a id="atom-selection" class="anchor" href="#atom-selection">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Atom selection</h3>
<p>Unlike most other libraries for computational chemistry, there is no text-based language to select a subset of atoms (for now).
However, one can achieve a rather similar experience using Crystal's own syntax:</p>
<pre><code class="language-crystal">struc.atoms.<span class="k">select</span> { <span class="o">|</span>atom<span class="o">|</span> atom.partial_charge <span class="o">&gt;</span> <span class="n">0</span> }
<span class="c"># or (Crystal&#39;s short block syntax)</span>
struc.atoms.<span class="k">select</span> <span class="o">&amp;</span>.partial_charge.<span class="o">&gt;</span>(<span class="n">0</span>)
<span class="c"># compared to a custom language</span>
struc.atoms.<span class="k">select</span> <span class="s">&quot;partial_charge &gt; 0&quot;</span>

<span class="c"># select atoms within a cylinder of radius = 4 A along the Z axis and centered at the origin</span>
struc.atoms.<span class="k">select</span> { <span class="o">|</span>atom<span class="o">|</span> atom.x<span class="o">**</span><span class="n">2</span> <span class="o">+</span> atom.y<span class="o">**</span><span class="n">2</span> <span class="o">&lt;</span> <span class="n">4</span> }
<span class="c"># compared to a custom language</span>
struc.atoms.<span class="k">select</span> <span class="s">&quot;sqrt(x) + sqrt(y) &lt; 4&quot;</span> <span class="c"># or &quot;x**2 + y**2 &lt; 4&quot;</span></code></pre>
<p>Using Crystal itself for selection provides one big advantage: type-safety.
Doing something like <code>atom.name**2</code> will result in an error during <em>compilation</em>, pointing exactly the error's location.
Instead, using a custom language will produce an error during <em>runtime</em> at some point during execution, where the message may be useful or not depending on the type of error.</p>
<p>Additionally, the code block can be as big and complex as necessary with multiple intermediary computations.
Furthermore, a negative condition may be confusing and not be trivial to write, but in Crystal you would simply use the <code>#reject</code> method instead.
The same syntax can also be used for counting, grouping, etc. via the standard library.</p>
<p>Thanks to the topology hierarchical access, the above also works for chains and residues:</p>
<pre><code class="language-crystal"><span class="c"># select protein chains</span>
struc.chains.<span class="k">select</span> <span class="o">&amp;</span>.residue.any?(<span class="o">&amp;</span>.protein?)
<span class="c"># select only solvent residues</span>
struc.residues.<span class="k">select</span> <span class="o">&amp;</span>.solvent?
<span class="c"># select residues with its CA atom within 5 A of the first CA atom (this is equivalent to &quot;same residue as&quot; or &quot;fillres&quot; in other libraries)</span>
ca <span class="o">=</span> struc.dig(<span class="s">&#39;A&#39;</span>, <span class="n">1</span>, <span class="s">&quot;CA&quot;</span>)
struc.residues.<span class="k">select</span> { <span class="o">|</span>residue<span class="o">|</span> residue.dig(<span class="s">&quot;CA&quot;</span>).coords.distance ca.coords <span class="o">&lt;</span> <span class="n">5</span> }</code></pre>
<p>This may improve performance drastically when selecting chains/residues as it only requires traversing the chains/residues, which will be significantly smaller than the number of atoms.
Most libraries do not offer such functionality, and one often needs to resort to select unique atoms within the desired residues.</p>
<h3><a id="coordinates-manipulation" class="anchor" href="#coordinates-manipulation">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Coordinates manipulation</h3>
<p>All coordinates manipulation is done using a <code>CoordinatesProxy</code> instance, available for any topology object containing atoms (<em>i.e.</em> structure, chain, and residue) via the <code>#coords</code> method:</p>
<pre><code class="language-crystal"><span class="c"># geometric center</span>
struc.coords.center
<span class="c"># center at origin</span>
struc.coords.center_at_origin
<span class="c"># wraps atoms into the primary unit cell</span>
struc.coords.wrap
<span class="c"># rotate about an axis</span>
struc.coords.rotate <span class="t">Chem</span><span class="t">::</span><span class="t">Spatial</span><span class="t">::</span><span class="t">Vec3</span>[<span class="n">1</span>, <span class="n">2</span>, <span class="n">3</span>], <span class="n">90</span>.degrees
<span class="c"># align coordinates to a reference structure</span>
struc.coords.align_to ref_struc</code></pre>
<h2><a id="benchmark" class="anchor" href="#benchmark">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Benchmark</h2>
<p><code>chem.cr</code> is implemented in pure Crystal, making it as fast or even faster than
some C-powered packages.</p>
<p>There is a benchmark at the <a href="https://github.com/franciscoadasme/pdb-bench">pdb-bench</a> repository that compares <code>chem.cr</code> with popular software for computational chemistry.
The benchmark includes the following tests:</p>
<ul>
<li>Read and parse PDB files.</li>
<li>Counting residues matching a condition.</li>
<li>Calculating distances.</li>
<li>Calculating the Ramachandran angles.</li>
<li>Aligning two structures.</li>
</ul>
<p>Overall, <code>chem.cr</code> (orange) comes first in most tests, sometimes over two orders of magnitude faster than the tested software.
Otherwise, it is slightly slower than the faster software, even compared to C/C++ code.</p>
<p>Parsing a large PDB file like <code>1HTQ</code> seems to be slow in <code>chem.cr</code>, but the implementation may drastically differ between softwares (e.g. error checking, implicit bond guessing).
Please refer to the table at the <a href="https://github.com/franciscoadasme/pdb-bench#results">benchmark results</a> for a detailed comparison.</p>
<p><img src="https://github.com/franciscoadasme/pdb-bench/raw/master/assets/bench.png" alt="" /></p>
<h2><a id="roadmap" class="anchor" href="#roadmap">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Roadmap</h2>
<h3><a id="topology-manipulation" class="anchor" href="#topology-manipulation">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Topology manipulation</h3>
<ul>
<li>[x] Automatic connectivity detection (includes periodic systems).</li>
<li>[x] Automatic bond order assignment.</li>
<li>[x] Residue templates.
<ul>
<li>[x] Custom terminal groups.</li>
</ul>
</li>
<li>[x] Automatic topology reconstruction based on residue templates.</li>
<li>[x] Atom wildcards (<code>&quot;C*&quot;</code> will select <code>&quot;C&quot;</code>, <code>&quot;CA&quot;</code>, <code>&quot;CB&quot;</code>, etc.).</li>
<li>[ ] Atom selection language.</li>
</ul>
<h3><a id="input-and-output" class="anchor" href="#input-and-output">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Input and Output</h3>
<ul>
<li>[x] Automatic file format detection.</li>
<li>[x] Support for per-file format options (<em>e.g.</em> select PDB chain).</li>
<li>[x] Friendly errors (message with error location).</li>
<li>[x] Iterator-based IO.</li>
<li>[ ] Compressed files (.gz and .xz).</li>
<li>[x] Trajectory support (basic implementation).</li>
</ul>
<h4><a id="file-formats" class="anchor" href="#file-formats">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>File formats</h4>
<ul>
<li>[x] PDB (.pdb, .ent).</li>
<li>[ ] Macromolecular Crystallographic Information Framework (CIF).</li>
<li>[ ] MacroMolecular Transmission Format (MMTF).</li>
<li>[x] Extended XYZ (.xyz).</li>
<li>[x] Mol/SDF (.mol, .sdf).</li>
<li>[x] Tripos Mol2 (.mol2).</li>
<li>[x] PSF (.psf).</li>
<li>[ ] Maestro (.mae).</li>
<li>[x] DCD trajectory format (.dcd).</li>
<li>[x] JDFTx (.ionpos and .lattice).</li>
<li>[x] VASP's Poscar (POSCAR, CONTCAR).</li>
<li>[x] DFTB+'s Gen format (.gen).</li>
</ul>
<h3><a id="analysis" class="anchor" href="#analysis">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Analysis</h3>
<ul>
<li>[x] Coordinates manipulation.</li>
<li>[x] Spatial calculations (distance, angle, dihedral, quaternions, affine transformations).</li>
<li>[x] Periodic boundary conditions (PBC) support (topology-aware wrap and unwrap).</li>
<li>[x] Secondary structure assignment.
<ul>
<li>[x] DSSP (native implementation).</li>
<li>[x] STRIDE (uses external program for now).</li>
<li>[x] PSIQUE (own method).</li>
</ul>
</li>
<li>[x] Nearest neighbor search (via native k-d tree).</li>
<li>[x] RMSD.</li>
<li>[x] Structure superposition.</li>
<li>[ ] Intermolecular interactions (H-bonds, etc.).</li>
<li>[x] Volumetric data.</li>
<li>[ ] Parallel processing.</li>
</ul>
<h3><a id="other" class="anchor" href="#other">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Other</h3>
<ul>
<li>[ ] Documentation.</li>
<li>[ ] Guides.</li>
<li>[ ] Workflows (handle calculation pipelines).</li>
<li>[ ] More tests.</li>
</ul>
<h2><a id="similar-software" class="anchor" href="#similar-software">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Similar software</h2>
<p>There are several libraries providing similar functionality.
Here we list some of them and provide one or more reasons for why we didn't use them. However, each one of them is good in their own right, so please do check them out if <code>chem.cr</code> does not work for you.</p>
<ul>
<li><a href="http://chemfiles.org/">Chemfiles</a> is a library for reading and writing chemistry files written in C++, but can also be used from other languages such as Python.
It is mainly focused on simulation trajectories and does not provide an object-oriented topology access.
Also, it does not provide functionality beyond parsing and writing files.</li>
<li><a href="https://openbabel.org/wiki/Main_Page">OpenBabel</a> is a massive C++ library providing support for more than 110 formats.
Its API is very complex to use.</li>
<li><a href="http://mdtraj.org/latest/">MDTraj</a>, <a href="http://www.mdanalysis.org/">MDAnalysis</a>, <a href="https://cclib.github.io">cclib</a>, <a href="https://pymatgen.org">pymatgen</a>, <a href="http://prody.csb.pitt.edu">prody</a>, <a href="https://biopython.org">Biopython</a> and more are written in Python, which usually means two things.
First, no type safety makes them sometimes difficult to use if there is not enough documentation (more common than one may think).
Second, need to deal with C for performance critical code.</li>
<li><a href="http://www.ks.uiuc.edu/Research/vmd/">VMD</a> and <a href="https://schrodinger.com/">Schrodinger</a> are very popular software for molecular visualization that provide an API in Tcl and/or Python to manipulate molecules.
However, these usually suffer from poor documentation, and they are difficult to extend functionality.</li>
</ul>
<h2><a id="testing" class="anchor" href="#testing">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Testing</h2>
<p>Run the tests with the <code>crystal spec</code> command.
Some tests work by compiling small programs, which may be slower to run.
These may be skipped by running:</p>
<pre><code class="language-console">$ crystal spec --tag='~codegen'</code></pre>
<h2><a id="contributing" class="anchor" href="#contributing">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Contributing</h2>
<ol>
<li>Fork it (<a href="https://github.com/franciscoadasme/chem.cr/fork">https://github.com/franciscoadasme/chem.cr/fork</a>)</li>
<li>Create your feature branch (<code>git checkout -b my-new-feature</code>)</li>
<li>Commit your changes (<code>git commit -am 'Add some feature'</code>)</li>
<li>Push to the branch (<code>git push origin my-new-feature</code>)</li>
<li>Create a new Pull Request</li>
</ol>
<h2><a id="contributors" class="anchor" href="#contributors">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>Contributors</h2>
<ul>
<li><a href="https://github.com/franciscoadasme">franciscoadasme</a> Francisco Adasme -
creator, maintainer</li>
</ul>
<h2><a id="license" class="anchor" href="#license">
  <svg class="octicon-link" aria-hidden="true">
    <use href="#octicon-link"/>
  </svg>
</a>License</h2>
<p>Licensed under the MIT license, see the separate LICENSE file.</p>
</div>
</body>
</html>
